002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,256 @@
+
+| Into            | stringr                 | tidyr               |
+|-----------------|-------------------------|---------------------|
+| Named/fixed     | `str_separate_wider()`  | `separate_wider()`  |
+| Unnamed/varying | `str_separate_longer()` | `separate_longer()` |
+
+Then each function has a `split` argument which can be one of the following cases:
+
+-   `by_position(1, 2, 3)`
+-   `by_width(1, 2, 3)`
+-   `by_separator(",", ignore.case = TRUE)`
+-   `by_regex_group("a(.)c")` (`by_group()` is too short/ambiguous).
+
+(This could also be defined by 3-4 mutual exclusive arguments, but that is hard to extend and hard to document)
+
+Pros:
+
+-   8 functions instead of 10, and extending with more splitting types is easy.
+-   Composing functions in this way seems more "tidyverse" like.
+
+Cons:
+
+-   Composing the functions on your own makes it harder to see the full set of options, and there's not enough total options to make the flexibility really worth it.
+-   Will be tempting to use `someotherfunction(x, by_position(1, 2, 3))`.
+
+Overall, this proposal only reduces the total number of functions by two, in return for a small gain.
+
+### Eliminate tidyr functions
+
+Can we eliminate the tidyr wrappers altogether and just use dplyr + stringr?
+This seemed like the "obvious" solution to me at least twice, so I wanted to make sure to explain why it doesn't work.
+
+To make the discussion concrete I'll start with quick and dirty implementations of `str_separate_by_wider()` and `str_separate_by_longer()`:
+
+```{r}
+str_separate_by_wider <- function(x, pattern, into) {
+  mat <- str_split_fixed(x, pattern, length(into))
+  colnames(mat) <- into
+  tibble::as_tibble(mat)
+}
+str_separate_by_longer <- function(x, pattern) {
+  str_split(x, pattern)
+}
+
+str_separate_by_wider(c("a-b", "c-d"), "-", c("first", "second"))
+str_separate_by_longer(c("a-b", "c-d-e"), "-")
+```
+
+We can use `str_separate_by_wider()` directly with `mutate()` to achieve the same affect as `separate()`:
+
+```{r}
+library(dplyr, warn.conflicts = FALSE)
+
+df <- tibble(x = c("a-b", "c-d"))
+df %>% separate(x, c("first", "second"), "-")
+
+df %>% 
+  mutate(
+    str_separate_by_wider(x, "-", c("first", "second")), 
+    .keep = "unused"
+  )
+```
+
+This works because `mutate()` will create multiple columns if an unnamed calculation returns a data frame.
+
+However, there's no straightforward way to use `str_separate_by_longer()` to create multiple columns.
+The simplest pure dplyr approach is probably to use `rowwise():`
+
+```{r}
+df %>%
+  rowwise() %>% 
+  summarise(str_separate_by_longer(x, "-")[[1]])
+```
+
+But it's quite hard to explain why this works, and the use of `summarise()` means that you need to explicitly list any variables that you want to leave as is.
+This possibly speaks to a [missing verb](https://github.com/tidyverse/dplyr/issues/5874) in dplyr but it's also possible that creating rows is fundamentally a tidyr operation.
```

Comment:
> it's also possible that creating rows is fundamentally a tidyr operation.

I feel like this is correct, sort of. I think it has to do with the fact that tidyr operations are good at taking a single column and either:
- Widening it to make more columns
- Lengthening it to make more rows (and this is what is useful here)